dkc/core/models/quota.py

```diff
@@ -1,26 +1,62 @@
 from typing import Type
 
+from django.conf import settings
 from django.contrib.auth.models import User
-from django.db import models
+from django.core.exceptions import ValidationError
+from django.db import IntegrityError, models, transaction
 from django.db.models.signals import post_save
 from django.dispatch import receiver
-from girder_utils.models import SelectRelatedManager
+
+
+def _default_user_quota() -> int:
+    return settings.DKC_DEFAULT_QUOTA
 
 
 class Quota(models.Model):
-    allocation = models.BigIntegerField(default=0)
-    user = models.OneToOneField(User, null=True, on_delete=models.CASCADE)
-    # root_set
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                check=models.Q(used__lte=models.F('allowed')),
+                name='used_lte_allowed',
+            )
+        ]
+
+    used = models.PositiveBigIntegerField(default=0)
+    allowed = models.PositiveBigIntegerField(default=_default_user_quota)
+
+    # OneToOneField ensures that only one Quota per User exists
+    user = models.OneToOneField(User, on_delete=models.CASCADE, related_name='quota')
```

Comment:
Eventually, with "project-based" quotas, this can be nullable.

Ideally, the relationship would point the other way, but since `User` is practically not directly extensible, this is a reasonable substitute. Given the `_create_user_quota` signal handler, we can rely on `user.quota` always existing for now.

Reply:
Could we pull this out of here and have a separate associative table to map users to quotas? Then, every user could have an entry in that table, but not every quota would have to have an entry there (i.e. for any non-user associated quotas).

Reply:
I think that would be a `Profile` model ([Option 2 here](https://simpleisbetterthancomplex.com/tutorial/2016/07/22/how-to-extend-django-user-model.html)), which would also store any other user-related info that we wanted to extend.

Do you foresee any upcoming features where we want to store additional information (beyond just foreign key relationships) with the user? If so, I think it's worth a detour to add the `Profile` now.

If not, I think it's worth merging this approach as-is, to start giving us working quota support. As we extend to non-user-associated ("project") quotas, we can rework the implementation internally.

Reply:
Let's add the profile model, it's not much code and should be straightforward. Just make it a commit on this same topic.

Reply:
I started to create a `Profile` model. This is something that's supported by Factory Boy: https://factoryboy.readthedocs.io/en/latest/recipes.html#example-django-s-profile , but a significant complication is the fact that we'll have to annotate all of our factories (not just `UserFactory` and `ProfileFactory`) with `@factory.django.mute_signals(post_save)`, which will disable any `post_save` handlers during essentially all tests. Eventually, we might need this, but it will introduce some drag.

IMO, the current implementation is adequate. At runtime, the reverse relation `user.quota` is adequate to support this feature and any lookup failures will immediately log 500 errors. If we can't depend on signals to support some behavior (instead requiring DB-level guarantees), I think we'd be giving up an important measure of power that Django provides.

In some ways, attaching the `ForeignKey` to `Quota` is superior, particularly once that field is nullable (to support project quotas) in the future, since we'll be able to easily filter user quotas from project quotas (I'm not seeing how to do this if the `ForeignKey` reference points *to* `Quota`).

Given the downside of suppressing all `post_save` signals, I think we should only proceed with `Profile` if we are going to need it to store additional information on each user. Otherwise, I'd advocate to merge this as-is, and proceed with developing more advanced features on top.

Reply:
I just want to be clear that putting the user field here is a relational hack. We're misusing nullability to model what should be a relation. We can merge this as is, but I'd like to create an issue to address this. Insofar as this repository is meant to model best practices, one of the things I really want us to exemplify is proper use of relational databases with Django, which means we'll need to work through the warts. As nice as FactoryBoy is, we can't let it be the library driving our design choices.

Let's make the ticket, I'll take a crack at this myself later on.